br/pkg/streamhelper/advancer.go

```diff
@@ -323,33 +366,57 @@ func (c *CheckpointAdvancer) onTaskEvent(ctx context.Context, e TaskEvent) error
 	return nil
 }
 
-func (c *CheckpointAdvancer) setCheckpoint(cp uint64) bool {
-	if cp < c.lastCheckpoint {
+func (c *CheckpointAdvancer) setCheckpoint(ctx context.Context, s spans.Valued) bool {
+	cp := newCheckpointWithSpan(s)
+	if cp.TS < c.lastCheckpoint.TS {
 		log.Warn("failed to update global checkpoint: stale",
-			zap.Uint64("old", c.lastCheckpoint), zap.Uint64("new", cp))
+			zap.Uint64("old", c.lastCheckpoint.TS), zap.Uint64("new", cp.TS))
+		return false
+	}
+	// lastCheckpoint is not increased too long enough.
+	// assume the cluster has expired locks for whatever reasons.
+	if c.lastCheckpoint.needResolveLocks() {
```

Comment:
Should it be here? this will hold the mutex of the advancer.

Reply:
if resolve locks gets stuck and it has locks, the checkpoint still won't be able to move forward?

Reply:
If there are too many messages blocked in the channel, we may consume more memory.

Reply:
ok, then how about starts a goroutine in CheckpointAdvancer and check lastCheckpoint every minute?

Reply:
I think that would be nice.